### ISSUE TITLE
Some questions about GFS

### ISSUE THREAD

**Issue by mcgibbon** [CliMT/climt#49]
As I'm looking to refactor GFS to fit the new Sympl version, a few questions are coming up.

1) What should GFS really be? Is it an Implicit or a TimeStepper? Right now I'm leaning towards an Implicit for certain reasons. Specifically, you cannot supply GFS with an arbitrary Prognostic that computes a tendency for an arbitrary quantity and expect it will time step it. GFS will only step forward specific quantities as well as tracers.

2) It seems to me that it is not actually necessary for those four quantities to be tracers in GFS. There is some Fortran code that keeps track of which index is which quantity. But as far as I can tell those indices are never used in GFS, which makes sense, since they're passive tracers. When I refactor, I plan to refactor the code so no tracers are mandatory.

3) There are [lines](https://github.com/CliMT/climt/blob/develop/climt/_components/gfs/component.py#L339)

```
        raw_output_arrays['air_pressure_on_interface_levels'][:] = \
            raw_output_arrays['air_pressure_on_interface_levels'][:, :, ::-1]
```

which flip the vertical ordering of `air_pressure_on_interface_levels`, but similar lines do not flip the vertical ordering for any other variables. What vertical ordering does GFS require, and is it different for different variables?

4) What is the purpose of the following:

```
        mylist = ['air_pressure']

        for quantity in mylist:
            raw_output_arrays[quantity][:] = raw_input_arrays[quantity][:]

        [...]

        for quantity in mylist:
            raw_input_arrays[quantity][:] = raw_output_arrays[quantity][:]

**Comment by mcgibbon** [comment]
For 1), there's also the matter that I don't see where `self.prognostics` is set. The init signature has no place for prognostics to be passed.

Should we use sequential tendency updates so that the spectral dynamical core can evolve only dynamics, and then we can evolve other quantities using a separate time stepper in physical space?

**Comment by JoyMonteiro** [comment]
1) My worry about making GFS an `Implicit` is that it cannot be used in conjunction with an additional `TimeStepper` which steps the prognostics forward in time.  In general, I'm not happy assigning `Prognostic`s to an `Implicit` component. 

I would prefer GFS just error if it is supplied with a `Prognostic` which provides a tendency for a quantity which it does not step forward in time.

`self.prognostics` is set after GFS initialisation. The reason for this behaviour was to allow changing the physics used by GFS at any time. For instance, you might want to use some physics only before/after spin up, and this option allowed doing such a thing without having to restart the model.

No matter what strategy you use (wrt sequential tendency updates), the main issue remains -- eventually you will need to transform between physical and spectral spaces. If this is done at every timestep, we lose too much information and the dynamics does not evolve. Do you see a way around this?

2) Certainly all four tracers are not required. However, GFS keeps track of the first tracer, which is water vapour and is part of the dynamical equations. So, that tracer is a required input.

3) GFS only requires surface pressure and sigma levels, and it outputs 3D pressure field on mid and interface levels. for some strange reason, the calculation of pressure on interface levels in fortran flips the dimensions. Hence this code.

4) That is a hack to get GFS to work. There is a strange bug (that I never figured out the reason for) which would cause GFS to fail basic dynamics tests (DCMIP) if this line was commented out. I would suggest leaving it in and opening an issue to remind me to fix it if possible.

**Comment by mcgibbon** [comment]
1. `self.prognostics` should be set at init to match the `TimeStepper` API. If you need to change the physics during runtime, you can have an additional helper method to do so. Is this possible?

To call the physical parameterizations, things need to be in physical spaces, so I don't see why not step forward in physical space and move back to spectral space rather than transform the tendencies to spectral space and then step forward in spectral space. I would think these two options give numerically identical results. In other words, I don't understand how stepping the physics separately leads to needing additional transforms or additional numerical error. Could you explain this to me?

2. Ahg, that is really annoying and doesn't make much sense to me. If it's part of the dynamical equations then it's not a tracer and shouldn't be treated as one by GFS. I'll think about a way for us to work with this, though.

3. OK, so GFS requires pressure on interface levels to start at the top of the domain, but all other variables start at the bottom. I'll have the component to reflect this when I likely add the ordering properties to Sympl.

4. OK, I'll try to figure it out as I refactor the GFS code but if not I'll create an issue.

**Comment by JoyMonteiro** [comment]
1. Yes, I'm fine with setting the prognostics initially and then modifying them later.
The semi-implicit part of the time stepping in GFS is done in spectral space, because the
equations in vorticity-divergence form have a simple description in spectral space. Because of this,
the overall time-stepping has to be done in spectral space.

If we want to step forward in physical space, then after the semi-implicit part we need to transform
back to physical space. I had chosen to transform to physical space initially, but it failed because of the reasons I mentioned, that the transform kills small scale variability which is important for the general circulation.

If you see the python wrapper, it still allows for the arrays to be modified in physical space -- if it detects that they have been modified, then the spectral arrays are updated. However, if this is done
every time step, the above issue arises.

2. The equations for stepping the water vapour forward in time are identical to any other tracer. I think this is the reason it is being considered as a tracer.

**Comment by mcgibbon** [comment]
1. I'm still confused, could you give me a step-by-step description of what happens in terms of components being called and transforms? For example, something like start with physical and spectral components -> calculate tendencies in physical space -> convert tendencies to spectral space -> step forward in spectral space? Actually, is that the correct ordering? It seems like it should be, and has the advantage of keeping all dynamical updates in spectral space and never converting them to physical space.

This is a good example of somewhere that has specialized requirements in order to get a particular component to work. A model using the GFS dynamical core can't have an arbitrary sequential tendency update main loop and expect to still work, for numerical/scientific reasons. In particular, users should want those quantities stepped by the dynamical core to only be stepped by the dynamical core, and not updated in any other diagnostic or prognostic manner.

Because we basically want all Prognostic objects to be given to GFS with no other TimeStepper active, it should accept as wide a range of Prognostic objects as possible, and we do not want to exclude those that compute tendencies for other quantities. We can handle this by taking those tendencies which the Fortran code does not handle and stepping them separately. We also need to handle any diagnostic output from those Prognostic objects. Requiring that the diagnostic outputs are not anything we're stepping spectrally (u, v, water vapor, etc.) shouldn't be too bad of a requirement.

**Comment by JoyMonteiro** [comment]
Yes, your ordering is correct. Because the Laplacian is multiplication in spectral space, this makes semi-implicit time-stepping easy in spectral space, hence all model quantities are represented spectrally. The physical space quantities returned by GFS (python wrapper) are simply the transformed versions of the spectral arrays, which are not available to the user.

Yes, this is where `Federation` can build in some heuristics when it encounters a component like GFS. This is also the reason why I preferred GFS to be a subclass of `TimeStepper` than an instance.

I'm fine with the behaviour you propose in the last paragraph. I think the dynamics is 3rd order RK; so a 3rd order timestepping of other quantities should suffice.
